wip-0003.md

```diff
@@ -0,0 +1,129 @@
+
+BLS signatures allow us to aggregate signatures provided by a set of group members. In the case of Witnet, those members will be the ARS, as the consensus of the Witnet chain is highly dependant on the reputation of each fo the peers. The overall aggregated signature can be validated in the target chain in a single verification. However, a BLS verification is approximately 10 times more costly than a regular ECDSA verification. Thus, BLS gives us a big advantage for bridging with other chains, but little advantage when using it to sign transactions internally in the _Witnet network_.
+
+In consequence, we need to derive a way to make both BLS and ECDSA algorithms co-exist in Witnet. This proposal introduces different solutions to such problem.
+
+## Requirements
+As already said before, we need to maintain a state about the ARS in the block headers that future blocks can be validated upon. Such information will be stored as a merkle root of all the identities and their position in the ARS.
+
+However, note that the reputation is today associated to the PKH of the Secp256K1 public key. However, ARS members in the ARS merkle root should be identifiable by their BLS public key, as this will be used to verify their aggregated signature.
+
+Additionally, UTXOs are also associated with the PKH of the Secp256k1 public key. Given the clear overhead of implementing a BLS verification compared to a ECDSA verification, UTXOs should keep being spendable by secp256k1 private key holders.
+
+## Proposal: A mapping between Secp256 and BN256 public keys.
+
+### Specification
+Our first proposal is to make every node store a mapping with the relationship between a *Secp256k1* and the *altBN_128* key.
+
+#### Information storage
+
+**`secp256k1AltbnMap` mapping.** A new mapping to be stored in memory that is updated every time a new key is proposed. This mapping does not need to be stored on-chain, but can be constructed by information that is on-chain. This can be implemented can be extended to the struct TotalReputationSet:
+
+```rust
+// Everything here is pseudo-Rust, it's obviously not guaranteed to compile
+struct TotalReputationSet {
```

Comment:
```suggestion
struct ActiveReputationSet {
```

The ARS is mentioned earlier, so why do we use the TRS here?

Reply:
AFAIK the ARS contains the active identities, but the isnt the information of the reputation hold by each identity stored in totalreputationset? I might be wrong here, it would be good if we can clarify a bit more. Since the key is information associated with each pkh, I thought it should be in the same place where the reputation of each identity is stored

Reply:
There is a small problem with storing the public key in the TRS: identities with null reputation aren't stored there. So for example an identity that mines a block is counted as "active", but has no extra reputation.

We can change the behavior of the TRS to include active identities, or we can store this map in a separate structure.

Reply:
Interesting, lets use a different structure then. However that means that one should include the BLS key also when proposing a block correct?

Reply:
Yes, let's just include the BLS key with the VRF proof, since that's the only way to enter the ARS.

Reply:
Great, agree